### ISSUE TITLE
Context propagation is not invoked when a workflow start and is configured using WorkflowClientOptions

### ISSUE THREAD

**Issue by jrpedrianes** [temporalio/sdk-java#796]
## Expected Behavior
When starting a workflow, context propagated configured at workflow client level must be used.

## Actual Behavior
Is not invoked and must be defined using `WorkflowOptions`

## Steps to Reproduce the Problem

In the following link, I created an example class in which we can see that the "serializeContext" method is not invoked before the workflow is started.

https://gist.github.com/jrpedrianes/2759391fa22311edf214932eeb3fd4b3

## Specifications

  - Version: sdk-java 1.3.1

**Comment by Spikhalskiy** [comment]
Hi Jorge, it's like that by design.

1. Imagine your client and worker have different classes in their classpath. A context propagator that is registered on the client side could even not exist in a classpath of the worker.
2. Imagine we don't have the first problem and they do exist. ContextPropagator could have a non-default constructor, have external instances injected into it, etc. We don't require ContextPropagators to be serializable and this would impose serious limitations on them. So, it's not like we can safely serialize ContextPropagator and send it to the server.
3. Think about a client that is implemented in Go and worker code is in java, how do we propagate Go context propagator implementation to JVM worker?

We may one day consider adding functionality that a worker will not apply worker propagators if they were not being enabled on the client. But we definitely can't just set up propagators on a client-side.

**Comment by jrpedrianes** [comment]
Yes, I understand that we need to register the "same" ContextPropagators in all workers that are required, and must be implemented using different languages (Java, Go...)

But maybe my confusion was thinking that when I register a context propagator a WorkflowClient level, ie: 

```
        WorkflowClient client = WorkflowClient
            .newInstance(service, WorkflowClientOptions.newBuilder()
                .setContextPropagators(List.of(new DummyContextPropagator()))
                .validateAndBuildWithDefaults());
```

The context propagators register there, will be used in all the resources using that client, ie:
```
WorkerFactory factory = WorkerFactory.newInstance(client);
```
and:
```
        GreetingWorkflow workflow =
            client.newWorkflowStub(
                GreetingWorkflow.class,
                WorkflowOptions.newBuilder()
                    .setWorkflowId(WORKFLOW_ID)
                    .setTaskQueue(TASK_QUEUE)
                    .validateBuildWithDefaults());
```

Maybe it is useful to add a Javadoc documentation that the context propagator will not be applied when we are creating a client stub.

**Comment by Spikhalskiy** [comment]
Oh, I got what you mean now. If in the snippets above context propagators are not used in the stubs created this way, it's a bug.

**Comment by jrpedrianes** [comment]
I was debugging a workflow invocation and I could not see where the context propagation defined using  WorkflowClientOptions could be used.

Seems that should be used in: 

- [WorkflowStubImpl.java](https://github.com/temporalio/sdk-java/blob/master/temporal-sdk/src/main/java/io/temporal/internal/sync/WorkflowStubImpl.java)
- [RootWorkflowClientInvoker.java](https://github.com/temporalio/sdk-java/blob/master/temporal-sdk/src/main/java/io/temporal/internal/client/RootWorkflowClientInvoker.java)
- [RootWorkflowClientHelper.java](https://github.com/temporalio/sdk-java/blob/0fedb5438bac22445e0c3f660af3cd717b22665a/temporal-sdk/src/main/java/io/temporal/internal/client/RootWorkflowClientHelper.java)

But I could not see any reference to `clientOptions.getContextPropagators()`

**Comment by Spikhalskiy** [comment]
For now, as a workaround, set the propagators on the worker itself and the stubs. It should be fixed in 1.4.0 or 1.5.0.

**Comment by jrpedrianes** [comment]
Now, we are using an interceptor to add it to all the client stubs, so we only need to touch one class. 

Thank a lot!!
